lib/pbench/test/unit/server/test_user_auth.py

```diff
@@ -178,6 +183,49 @@ def test_user_relogin(client, server_config):
                 json={"username": "user", "password": "12345"},
             )
             assert response.status_code == HTTPStatus.OK
+            third_auth_token = response.json["auth_token"]
+            assert (
+                first_auth_token != third_auth_token
+            ), "Third login returned first token"
+            assert (
+                second_auth_token != third_auth_token
+            ), "Third login returned second token"
+
+    @staticmethod
+    def test_login_removes_expired_tokens(client, server_config):
+        """Test to ensure on a new login previously expired tokens are
+        removed."""
+        with client:
+            # user registration
+            resp_register = register_user(
+                client,
+                server_config,
+                username="me",
+                firstname="first name",
+                lastname="last name",
+                email="me@domain.com",
+                password="12345",
+            )
+            assert resp_register.status_code == HTTPStatus.CREATED
+
+            user = User.query(username="me")
+            # Add 3 expired tokens, values are in "days" old
+            for age in (1, 2, 4):
+                token = ActiveTokens(
+                    f"token-that-is-{age:d}-days-old",
+                    datetime.now(timezone.utc) - timedelta(days=age),
+                )
+                user.add_token(token)
+
+            response = client.post(
+                f"{server_config.rest_uri}/login",
+                json={"username": "me", "password": "12345"},
+            )
+            assert response.status_code == HTTPStatus.OK
+
+            for age in (1, 2, 4):
+                token = ActiveTokens.query(f"token-that-is-{age:d}-days-old")
+                assert token is None
```

Comment:
We should probably do a query before the login which shows that the tokens that we added are still there...otherwise, the check at line 228 doesn't prove much....

Reply:
But we just added them at line 218.

Reply:
We _requested_ that they be added at line 218, but we don't have any proof that the request had any effect...and then we believe that getting an empty result at line 227 means that the tokens that we "added" were actually removed.

As the code stands, if the assertion at line 228 _doesn't_ fail, there's an alternative explanation:  if could be that the tokens were simply never added at line 218....

Reply:
Actually, we need to commit that transaction.

Reply:
How could they "simply never be added at line 218"?  That is what `user.add_token()` method is supposed to do.  Why would this code, as opposed to any other code that calls the same method, need to verify they were added?

This is testing user auth code flow, not the database.  If the SQLAlchemy ORM fails to commit those rows to the database, we have a much larger problem.

Reply:
This is unit test code...it runs in a weird environment in which reality is regularly warped.

The assertion at line 228 is supposed to detect that the items have been removed from the database...why not perform the inverse check before invoking the CUT to ensure that the expected initial state is in place?  This ensures that the transition that we're trying to observe actually takes place.  ("Now we see them, now we don't -- the CUT is working.")

I don't think that it's a big deal to add the assertion (or that it will be costly to execute).  You evidently do, so I'll defer to you and close this conversation.